### PR TITLE
Upgrade building

### PR DESCRIPTION
**Pull request by pfeningviola**
Please, check the fantastic new feature! :-)
- upgrade the building
- show error messages, if the upgrade and the new button is inactive
- update the gold amount instantly (before the http response arrives) by buying or upgrading a building

### CODE REVIEW

File: Tribes/src/app/_components/building-type/building-type.component.ts [hunk 1]
```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
...
 
   checkIfBuildingIsUpgradeable(level: number): boolean {
-    return this.getNumberOfBuildingsByLevel(level) !== 0 && this.townhallLevel > level && this.goldAmount >= this.getUpgradingCost(level);
+    return this.getNumberOfBuildingsByLevel(level) !== 0
+            && this.townhallLevel > level
+            && this.goldAmount >= this.getUpgradingCost(level)
+            && !this.checkIfAllBuildingByLevelIsUnderConstruction(level);
+  }
+
+  checkUpgradingConditions(level: number): string {
```
Comment (symunona) [inline Tribes/src/app/_components/building-type/building-type.component.ts]: As a naming convention, I would rather use something like 
`isUpgradePossible`, which returns a boolean, or a `validateIfUpgradeBuildingIsPossible`. I am also not satisfied fully with these. Also, I would consider putting them into a service maybe? i.e. imagine if the requirement would be to "yeah, let's put the upgrade button out in another screen" - this logic can be isolated quite well.

File: Tribes/src/app/_components/building-type/building-type.component.ts [hunk 2]
```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
...
+        return 'The townhall level is too low.';
+      } else if (this.goldAmount < this.getUpgradingCost(level)) {
+        return 'You don\'t have enough money';
+      }
+    }
+  }
+
+  upgradeBuilding(level: number): void {
+    if (this.type === 'townhall' && !this.checkIfTownhallIsUpgrading()) {
+      if (this.goldAmount >= this.getUpgradingCost(level)) {
```
Comment (symunona) [inline Tribes/src/app/_components/building-type/building-type.component.ts]: Here, that function could be reused i.e. `isUpgradePossible`

File: Tribes/src/app/_components/building-type/building-type.component.ts [hunk 3]
```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
...
+        const idToUpgrade: number = this.buildings.find(building => building.type === 'TOWNHALL').id;
+        this.goldAmount = this.goldAmount - this.getUpgradingCost(level);
+        this.buildingService.upgradeBuilding(idToUpgrade, level + 1);
+      } else {
+        this.errorMessage = this.checkTownhallUpgradingConditions(level);
+      }
+    } else {
+      if (this.checkIfBuildingIsUpgradeable) {
```
Comment (symunona) [inline Tribes/src/app/_components/building-type/building-type.component.ts]: Is this supposed to be a function call?

File: Tribes/src/app/_components/building-type/building-type.component.ts [hunk 4]
```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
...
+      } else {
+        this.errorMessage = this.checkUpgradingConditions(level);
+      }
+    }
+  }
+
+  checkIfAllBuildingByLevelIsUnderConstruction(level: number): boolean {
+    const readyBuildings: Building[] = this.buildings.filter(building => !this.buildingService.checkIfBuildingIsProgressing(building));
+    return readyBuildings.length === 0;
+  }
```
Comment (symunona) [inline Tribes/src/app/_components/building-type/building-type.component.ts]: I am not sure if we have decided whether you can be upgrading multiple levels at the same time. I assume that this lets you be able to upgrade 1 per level? It is good, just did we agree on this?

File: Tribes/src/app/_components/building-type/building-type.component.ts [hunk 5]
```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
...
+
+  checkIfTownhallIsUpgrading(): boolean {
+    const townhall: Building = this.buildings.find(building => building.type === 'TOWNHALL');
+    console.log(townhall.finishedAt);
+    console.log(new Date().getTime());
```
Comment (symunona) [inline Tribes/src/app/_components/building-type/building-type.component.ts]: As a clean coding principle, we are not leaving console.log statements in code.

File: Tribes/src/app/_components/building-type/building-type.component.ts (hunk 4 above)
Comment (pfeningviola) [reply Tribes/src/app/_components/building-type/building-type.component.ts]: This condition check, if the user have any building of a level, which is not under construction not upgrading  at the moment. Only this case can upgrade one more building. We can have the situation, that all buildings of a level are upgrading, but they have the previous level until the process is finished, without this check we could begin the upgrading process of the same building twice.

File: Tribes/src/app/_components/building-type/building-type.component.ts (hunk 2 above)
Comment (pfeningviola) [reply Tribes/src/app/_components/building-type/building-type.component.ts]: By the townhall and by other buidlings the conditions to upgrade are not the same, therefore I think, that I can't use the same function.

File: Tribes/src/app/_components/building-type/building-type.component.ts (hunk 1 above)
Comment (pfeningviola) [reply Tribes/src/app/_components/building-type/building-type.component.ts]: I have 2 functions: 
1. checkIfTownhallIsUpgrading() -> renamed to isUpgradePossible(): this returns a boolean
2. checkUpgradingConditions() returns a string, that I use by creating the error message, I renamed it to validateUpgradingConditions() Is it better?

I was thinking to put all this functions into the building.service, but to do this, I need to put a lot of other functions to the service as well and I should use a lot of parameters (like townhallLevel, type, goldAmount), because the service doesn't know these. Should I put it into the service even so?

File: Tribes/src/app/_services/building.service.ts [hunk 6]
```diff
@@ -27,14 +28,47 @@ export class BuildingService {
...
+  checkIfBuildingIsProgressing(building: Building): boolean {
+    const currentTime = new Date().getTime();
+    return currentTime <= building.finishedAt;
+  }
+
+  setTimeoutsAgain() {
+    if (localStorage.length > 1) {
+      for (let i = 0; i < localStorage.length; i++) {
+        let key: string = localStorage.key(i);
+        if (key !== 'TOKEN') {
```
Comment (symunona) [inline Tribes/src/app/_services/building.service.ts]: You most definitely do not have to iterate over the items :)

You can just do something like this:
```
// To save:
const MY_KEY = 'someprefix_is_illdomos'
localStorage.setItem(MY_KEY, JSON.stringify(['timeout1', 'timeout2', ...]))
// To load:
let timeoutList = []
if (localStorage.getItem(MY_KEY)){
timeOutList = JSON.parse(localStorage.getItem('app_timeouts'))}
```

File: Tribes/src/app/_components/resources/resources.component.ts [hunk 7]
```diff
@@ -24,13 +25,25 @@ export class ResourcesComponent implements OnInit {
           this.getResources();
       }
     });
+    this.buildingService.updateRessourceByConstruction.subscribe({
+      next: (sum) => {
+        this.updateRessourceAfterConstruction(sum);
```
Comment (symunona) [inline Tribes/src/app/_components/resources/resources.component.ts]: I would rename `sum` to something like `upgradeCost`.

File: Tribes/src/app/_services/building.service.ts [hunk 8]
```diff
@@ -27,14 +29,52 @@ export class BuildingService {
...
+      savedTimeouts.push(finishedAt);
+      localStorage.setItem(this.keyTimeouts, JSON.stringify(savedTimeouts));
+    }
+  }
+
+  setTimeoutsAgain() {
+    if (localStorage.getItem(this.keyTimeouts)) {
+      let savedTimeouts = [];
+      savedTimeouts = JSON.parse(localStorage.getItem(this.keyTimeouts)).filter(timeout => (timeout > new Date().getTime()));
+      savedTimeouts.forEach(timeout => { setTimeout(() => { this.finishConstruction.emit(); }, timeout - new Date().getTime()); });
```
Comment (symunona) [inline Tribes/src/app/_services/building.service.ts]: This looks good enough for now, if it works.

However I'd move the whole out to an external function like `setFinishConstructionTimeout` that gets the time as a parameter. By the way I just noticed, that you do not save which building had the timeout. I assume when this event happens, you just remove one randomly? Or what minds this? Why do you have this in the building service, and why not in the progress bar is checking this?

File: Tribes/src/app/_components/notifications/notifications.component.ts [hunk 9]
```diff
@@ -25,31 +26,33 @@ export class NotificationsComponent implements OnInit {
           this.generateListToDisplay();
       }
     });
+    this.buildingService.updateRessourceByConstruction.subscribe({
+      next: () => {
+        this.emptyMessage = undefined;
```
Comment (symunona) [inline Tribes/src/app/_components/notifications/notifications.component.ts]: Whooohoooo... I have not noticed this! Why is this needed?

Comment (pfeningviola) [reply Tribes/src/app/_components/notifications/notifications.component.ts]: This is a message, which is shown in the sidebar if we don't have any building under construction. But if we pressed the button to add a new one, or to upgrade, we don't want to see anymore this message.

File: Tribes/src/app/_services/building.service.ts (hunk 8 above)
Comment (pfeningviola) [reply Tribes/src/app/_services/building.service.ts]: It's not randomly, the function will remove the building, which is finished, it's enough to know the finishedAt property of the building, the id is not necessary. 
There is more components, that are subsrcibe to the finishConstruction event, and this function is called, when we buy a new building or upgrade one, and this functions belong to the building.service.
